File: android/core/src/main/java/uk/co/amlcurran/social/CalendarSource.java
package uk.co.amlcurran.social;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import uk.co.amlcurran.social.utils.SparseArray;

class CalendarSource {

    private final SparseArray<CalendarSlot> calendarItems;
    private final int daysSize;
    private final TimeCalculator timeCalculator;
    private final TimeRepository timeRepository;

    public CalendarSource(SparseArray<CalendarSlot> calendarItems, int daysSize, TimeCalculator timeCalculator, TimeRepository timeRepository) {
        this.calendarItems = calendarItems;
        this.daysSize = daysSize;
        this.timeCalculator = timeCalculator;
        this.timeRepository = timeRepository;
    }

    public int count() {
        return daysSize;
    }

    @Nullable
    public CalendarItem itemAt(int position) {
        CalendarSlot calendarSlot = calendarItems.get(position);
        if (calendarSlot == null || calendarSlot.isEmpty()) {
            Timestamp startTime = startOfTodayBlock(position);
            Timestamp endTime = endOfTodayBlock(position);
            return new EmptyCalendarItem(startTime, endTime);
        }
        return calendarSlot.firstItem();
    }

    @Nonnull
    private Timestamp startOfTodayBlock(int position) {
        return timeCalculator.startOfToday().plusDays(position).plusHoursOf(timeRepository.borderTimeStart());
    }

    @Nonnull
    private Timestamp endOfTodayBlock(int position) {
        return timeCalculator.startOfToday().plusDays(position).plusHoursOf(timeRepository.borderTimeEnd());
    }

    @Nonnull
    public CalendarSlot slotAt(int position) {
        return calendarItems.get(position, new CalendarSlot());
    }

    public boolean isEmptySlot(int position) {
        return slotAt(position).isEmpty();
    }
}

File: android/core/src/main/java/uk/co/amlcurran/social/CalendarSource.kt
package uk.co.amlcurran.social

import java.util.*

class CalendarSource(private val calendarItems: HashMap<Int, CalendarSlot>, private val daysSize: Int, private val timeCalculator: TimeCalculator, private val timeRepository: TimeRepository) {

    fun count(): Int {
        return daysSize
    }

    fun itemAt(position: Int): CalendarItem? {
        val calendarSlot = calendarItems.get(position)
        if (calendarSlot == null || calendarSlot.isEmpty) {
            val startTime = startOfTodayBlock(position)
            val endTime = endOfTodayBlock(position)
            return EmptyCalendarItem(startTime, endTime)
        }
        return calendarSlot.firstItem()
    }

    private fun startOfTodayBlock(position: Int): Timestamp {
        return timeCalculator.startOfToday().plusDays(position).plusHoursOf(timeRepository.borderTimeStart())
    }

    private fun endOfTodayBlock(position: Int): Timestamp {
        return timeCalculator.startOfToday().plusDays(position).plusHoursOf(timeRepository.borderTimeEnd())
    }

    fun slotAt(position: Int): CalendarSlot {
        return calendarItems[position] ?: CalendarSlot()
    }

    fun isEmptySlot(position: Int): Boolean {
        return slotAt(position).isEmpty
    }
}
